lib/streamlit/watcher/file_watcher.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2018-2021 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Union, Type, Callable
+
+from streamlit import config
+from streamlit import env_util
+from streamlit.logger import get_logger
+from streamlit.watcher.polling_file_watcher import PollingFileWatcher
+
+LOGGER = get_logger(__name__)
+
+try:
+    # Check if the watchdog module is installed.
+    from streamlit.watcher.event_based_file_watcher import EventBasedFileWatcher
+
+    watchdog_available = True
+except ImportError:
+    watchdog_available = False
+    if not config.get_option("global.disableWatchdogWarning"):
+        msg = "\n  $ xcode-select --install" if env_util.IS_DARWIN else ""
+
+        LOGGER.warning(
+            """
+  For better performance, install the Watchdog module:
+  %s
+  $ pip install watchdog
+
+        """
+            % msg
+        )
+
+FileWatcherType = Union[
+    Type[EventBasedFileWatcher],
+    Type[PollingFileWatcher],
+]
+
+
+def watch_file(path: str, on_file_changed: Callable[[str], None]) -> bool:
+    """Create a FileWatcher for the given file if we have a viable
+    FileWatcher class.
+
+    Parameters
+    ----------
+    path
+        Path of the file to watch.
+    on_file_changed
+        Function that's called when the file changes.
+
+    Returns
+    -------
+    bool
+        True if the file is being watched, or False if we have no
+        FileWatcher class.
+    """
+    watcher_class = get_file_watcher_class()
+    if watcher_class is None:
+        return False
+
+    watcher_class(path, on_file_changed)
+    return True
```

Comment:
Hm, I don't have a strong understanding of how the Python garbage collector works (so this may be a longshot), but could this cause any weirdness with the gc when using `PollingFileWatcher`?

The refcount on `EventBasedFileWatcher`s should always be >0 since there's an internal singleton that keeps track of all the individual watchers, but in the case of `PollingFileWatcher`, we seem to lose the reference to it, which would make it a candidate to be garbage collected and the `on_file_changed` callback mysteriously stop working.

Maybe returning a ref to an `Optional[FileWatcherType]` could be done here, but I'm not entirely sure if my concerns are valid to begin with.

Reply:
Whoops, never mind I realized what I was missing. The `ThreadPoolExecutor` in `PollingFileWatcher` has a ref to the instance (which, now that I'm thinking about it, should have been obvious since it needs a ref to the callback to run it)

Reply:
I added a comment to PollingFileWatcher to clarify that it's not necessary to retain an instance to have the polling work, because it's definitely non-obvious at first glance!